File: backend/apps/terminals/runs_effect_port.py
"""The temporary terminal executor behind the Slice 3 Runs handoff.

Rust owns every Runs table. This module is the whole of what Django may still
do about a launch: observe what exists under a deterministic runtime identity,
and create the runtime for one effect Rust has already made durable. It reads
its own terminals-owned launch request; it never writes an Agent Run,
Automation Attempt, Launch Effect, or Status Event row, and it receives no
attempt identity or authority to mint a run.
"""

from __future__ import annotations

import logging

from apps.runs.write_ownership import rust_owns_runs_writes
from apps.terminals.models import AgentTerminalSession, TerminalLaunchRequest
from apps.terminals.runtime import (
    CreateTerminal,
    TerminalAlreadyExists,
    TerminalObservationError,
    TerminalState,
)


logger = logging.getLogger(__name__)

PORT_VERSION = 1


def readiness() -> dict:
    """The exact record Rust requires before it will call this executor."""

    return {
        "version": PORT_VERSION,
        "ready": rust_owns_runs_writes(),
        "runs_owner": "rust",
        "effect_owner": "django",
        # Stated rather than implied: there is no Django Runs writer left to
        # fall back to, so a failed handoff cannot silently downgrade.
        "django_runs_write_fallback": False,
    }


def _runtime():
    """The live runtime seam. Imported lazily so this port stays importable
    without pulling the whole launch module into every consumer."""

    from apps.terminals.launch import terminal_runtime

    return terminal_runtime


def observe(identity: dict) -> dict:
    """Report what exists under one deterministic runtime identity.

    Only three answers permit a decision, and the honest fourth — "I could not
    tell" — is the default. Reconciliation reads ``absent`` as permission to
    execute again, so an unproven answer must never be reported as absent.
    """

    agent_run_id = str(identity.get("agent_run_id") or "")
    if not agent_run_id:
        return {"observation": "uncertain", "detail": "identity_missing"}
    try:
        observation = _runtime().inspect(agent_run_id)
    except TerminalObservationError:
        return {"observation": "uncertain", "detail": "runtime_unobservable"}
    except Exception:
        logger.warning("runs effect observation failed run=%s", agent_run_id, exc_info=True)
        return {"observation": "uncertain", "detail": "runtime_unobservable"}

    if observation.state is TerminalState.MISSING:
        return {"observation": "absent"}
    if observation.state is TerminalState.EXITED:
        # The runtime the effect asked for is gone. Its exit is a terminal fact
        # Rust records through the lifecycle seam, not something to adopt here.
        return {"observation": "absent"}

    mirror = AgentTerminalSession.objects.filter(agent_run_id=agent_run_id).first()
    if mirror is None:
        return {
            "observation": "conflicting",
            "runtime_id": agent_run_id,
            "detail": "runtime_without_terminal_mirror",
        }
    expected_issue = str(identity.get("issue_id") or "")
    request = TerminalLaunchRequest.objects.filter(agent_run_id=agent_run_id).first()
    if request is not None and expected_issue and request.issue_id != expected_issue:
        return {
            "observation": "conflicting",
            "runtime_id": agent_run_id,
            "detail": "runtime_identity_conflict",
        }
    return {"observation": "live", "runtime_id": agent_run_id}


def execute(claim: dict) -> dict:
    """Create the deterministic runtime for one already-durable effect.

    The claim carries exactly the two identities Rust predetermined. Everything
    else — the approved command, working directory, and environment — is read
    from this capability's own durable launch request, so no command line,
    path, or credential ever crosses the ownership boundary.
    """

    effect_id = str(claim.get("effect_id") or "")
    agent_run_id = str(claim.get("agent_run_id") or "")
    if not effect_id or not agent_run_id:
        return {"ok": False, "code": "claim_incomplete", "retryable": False,
                "cleanup_confirmed": True}

    request = TerminalLaunchRequest.objects.filter(
        effect_id=effect_id, agent_run_id=agent_run_id
    ).first()
    if request is None:
        # The effect is durable but its execution material is not. Nothing was
        # started, so cleanup is confirmed and a blind retry is pointless.
        return {"ok": False, "code": "launch_request_unavailable",
                "retryable": False, "cleanup_confirmed": True}

    try:
        adopted = _create_or_adopt(request)
    except Exception as exc:
        logger.exception("terminal effect execution failed run=%s", agent_run_id)
        return {
            "ok": False,
            "code": "terminal_runtime_unavailable",
            "retryable": True,
            # A failed create may have left a partial runtime behind, so
            # cleanup is confirmed only when termination proved it gone.
            "cleanup_confirmed": _cleanup_confirmed(agent_run_id),
            "detail": exc.__class__.__name__,
        }
    _record_terminal_mirror(request)
    return {"ok": True, "runtime_id": agent_run_id, "adopted": adopted}


def _create_or_adopt(request: TerminalLaunchRequest) -> bool:
    """Create the runtime, or adopt the one already holding this identity."""

    try:
        _runtime().create(
            CreateTerminal(
                agent_run_id=request.agent_run_id,
                command=request.command,
                working_directory=request.working_directory,
                environment=dict(request.environment or {}),
                dimensions=request.dimensions_tuple(),
            )
        )
        return False
    except TerminalAlreadyExists:
        # A crash between creation and acknowledgement leaves exactly this
        # state. Adoption is the idempotent success path; a second runtime is
        # never created for one durable effect.
        return True


def _record_terminal_mirror(request: TerminalLaunchRequest) -> None:
    """Record the terminals-owned mirror. This table is not Rust-owned."""

    AgentTerminalSession.objects.update_or_create(
        agent_run_id=request.agent_run_id,
        defaults={
            "tmux_session_name": request.agent_run_id,
            "task_id": request.task_id,
            "module_id": request.module_id,
            "project_id": request.project_id,
            "agent": request.agent,
            "created_at": request.created_at,
            "terminated_at": None,
            "runtime_namespace": _runtime().namespace,
            "runtime_cleanup_pending": False,
            "scope": request.scope,
            "doc_rel_path": request.doc_rel_path,
        },
    )


def _cleanup_confirmed(agent_run_id: str) -> bool:
    try:
        _runtime().terminate(agent_run_id)
        return True
    except Exception:
        logger.warning(
            "terminal effect cleanup unconfirmed run=%s", agent_run_id, exc_info=True
        )
        return False

File: studio/src-tauri/src/runs_effect_port.rs
//! The temporary compatibility boundary to the still-Python terminal
//! capability.
//!
//! Rust owns every Runs table. Django owns only the external terminal runtime,
//! so this port asks it two questions and nothing else: what exists under a
//! deterministic runtime identity, and please create the runtime for one
//! already-durable effect. Both directions carry identities and typed
//! outcomes; neither carries a prompt, a command line, a path, or a credential,
//! and Django is given no attempt identity or authority to mint a run.

use async_trait::async_trait;
use reqwest::{Client, Method};
use serde::Deserialize;
use serde_json::{json, Value};

use crate::runs_persistence::{
    ClaimedLaunch, LaunchExecutor, LaunchExecutorFailure, LaunchRuntimeEvidence,
    LaunchRuntimeProbe, RuntimeIdentity, RuntimeObservation,
};

const READINESS_PATH: &str = "/terminals/runs-effects";
const OBSERVE_PATH: &str = "/terminals/runs-effects/observe";
const EXECUTE_PATH: &str = "/terminals/runs-effects/execute";

/// The exact health record the Python executor must publish before Rust will
/// call it. Anything else keeps Slice 3 readiness closed.
#[derive(Debug, Deserialize, Eq, PartialEq)]
struct ExecutorReadiness {
    version: i32,
    ready: bool,
    runs_owner: String,
    effect_owner: String,
    django_runs_write_fallback: bool,
}

#[derive(Clone)]
pub struct RunsEffectPort {
    client: Client,
    base_url: String,
    api_key: String,
}

impl RunsEffectPort {
    pub fn new(base_url: String, api_key: String) -> Self {
        Self {
            client: Client::new(),
            base_url: base_url.trim_end_matches('/').to_owned(),
            api_key,
        }
    }

    /// Prove the temporary executor is present, healthy, and has itself given
    /// up every Runs writer. A missing, malformed, or partial answer is a
    /// closed gate, not a warning.
    pub async fn verify_health(&self) -> Result<(), String> {
        let value = self
            .request(Method::GET, READINESS_PATH, None)
            .await
            .map_err(|_| "The terminal Runs effect port is unavailable.".to_owned())?;
        let readiness: ExecutorReadiness = serde_json::from_value(value)
            .map_err(|_| "The terminal Runs effect port returned an invalid health record.")?;
        if readiness
            != (ExecutorReadiness {
                version: 1,
                ready: true,
                runs_owner: "rust".to_owned(),
                effect_owner: "django".to_owned(),
                django_runs_write_fallback: false,
            })
        {
            return Err(
                "The terminal Runs effect port has not given up Django Runs writes.".into(),
            );
        }
        Ok(())
    }

    async fn request(
        &self,
        method: Method,
        path: &str,
        body: Option<Value>,
    ) -> Result<Value, String> {
        let mut request = self
            .client
            .request(method, format!("{}{path}", self.base_url))
            .header("x-api-key", &self.api_key);
        if let Some(body) = body {
            request = request.json(&body);
        }
        let response = request
            .send()
            .await
            .map_err(|_| "effect_port_unavailable".to_owned())?;
        let status = response.status();
        let value = response
            .json::<Value>()
            .await
            .map_err(|_| "effect_port_invalid_response".to_owned())?;
        if status.is_success() {
            Ok(value)
        } else {
            Err(value
                .get("code")
                .or_else(|| value.get("detail"))
                .and_then(Value::as_str)
                .unwrap_or("effect_port_failed")
                .to_owned())
        }
    }
}

/// Exactly the immutable intent the executor needs to find or create the
/// deterministic runtime. Nothing here is secret or executable.
fn identity_payload(identity: &RuntimeIdentity) -> Value {
    json!({
        "effect_id": identity.effect_id,
        "agent_run_id": identity.agent_run_id,
        "project_id": identity.project_id,
        "issue_id": identity.issue_id,
        "scope": identity.scope,
        "provider": identity.provider,
        "target_kind": identity.target_kind,
        "target_id": identity.target_id,
    })
}

#[async_trait]
impl LaunchRuntimeProbe for RunsEffectPort {
    async fn observe(&self, identity: RuntimeIdentity) -> RuntimeObservation {
        let value = match self
            .request(
                Method::POST,
                OBSERVE_PATH,
                Some(identity_payload(&identity)),
            )
            .await
        {
            Ok(value) => value,
            // An unreachable executor proves nothing. Reconciliation must not
            // read that as "absent" and start a second terminal.
            Err(detail) => return RuntimeObservation::Uncertain { detail },
        };
        let runtime_id = value
            .get("runtime_id")
            .and_then(Value::as_str)
            .unwrap_or_default()
            .to_owned();
        match value.get("observation").and_then(Value::as_str) {
            Some("absent") => RuntimeObservation::Absent,
            Some("live") if !runtime_id.is_empty() => RuntimeObservation::Live { runtime_id },
            Some("conflicting") if !runtime_id.is_empty() => RuntimeObservation::Conflicting {
                runtime_id,
                detail: value
                    .get("detail")
                    .and_then(Value::as_str)
                    .unwrap_or("runtime_identity_conflict")
                    .to_owned(),
            },
            _ => RuntimeObservation::Uncertain {
                detail: "effect_port_unreadable_observation".to_owned(),
            },
        }
    }
}

#[async_trait]
impl LaunchExecutor for RunsEffectPort {
    async fn execute(
        &self,
        claim: ClaimedLaunch,
    ) -> Result<LaunchRuntimeEvidence, LaunchExecutorFailure> {
        // A claim carries exactly two identities. The executor resolves its own
        // provider and runtime policy from approved sources, so nothing else
        // needs to — or may — cross this boundary.
        let payload = json!({
            "effect_id": claim.effect_id,
            "agent_run_id": claim.agent_run_id,
        });
        let value = self
            .request(Method::POST, EXECUTE_PATH, Some(payload))
            .await
            .map_err(|code| LaunchExecutorFailure {
                code,
                message: "The terminal effect executor did not answer.".to_owned(),
                retryable: true,
                // An executor that never answered cannot have proven cleanup,
                // so the effect stays cleanup-pending rather than being closed.
                cleanup_confirmed: false,
            })?;
        if value.get("ok").and_then(Value::as_bool) == Some(true) {
            let runtime_id = value
                .get("runtime_id")
                .and_then(Value::as_str)
                .unwrap_or_default()
                .to_owned();
            if runtime_id.is_empty() {
                return Err(LaunchExecutorFailure {
                    code: "effect_port_invalid_response".to_owned(),
                    message: "The terminal effect executor reported no runtime.".to_owned(),
                    retryable: false,
                    cleanup_confirmed: false,
                });
            }
            return Ok(LaunchRuntimeEvidence {
                runtime_id,
                adopted: value.get("adopted").and_then(Value::as_bool) == Some(true),
            });
        }
        Err(LaunchExecutorFailure {
            code: value
                .get("code")
                .and_then(Value::as_str)
                .unwrap_or("effect_port_failed")
                .to_owned(),
            message: "The terminal effect executor could not create the runtime.".to_owned(),
            retryable: value.get("retryable").and_then(Value::as_bool) != Some(false),
            cleanup_confirmed: value.get("cleanup_confirmed").and_then(Value::as_bool)
                == Some(true),
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn only_the_exact_health_record_is_accepted() {
        let complete = json!({
            "version": 1,
            "ready": true,
            "runs_owner": "rust",
            "effect_owner": "django",
            "django_runs_write_fallback": false,
        });
        assert!(serde_json::from_value::<ExecutorReadiness>(complete.clone()).is_ok());

        for (field, value) in [
            ("ready", json!(false)),
            ("runs_owner", json!("django")),
            ("django_runs_write_fallback", json!(true)),
        ] {
            let mut degraded = complete.clone();
            degraded[field] = value;
            let readiness: ExecutorReadiness =
                serde_json::from_value(degraded).expect("decode the degraded record");
            assert_ne!(
                readiness,
                ExecutorReadiness {
                    version: 1,
                    ready: true,
                    runs_owner: "rust".to_owned(),
                    effect_owner: "django".to_owned(),
                    django_runs_write_fallback: false,
                }
            );
        }
    }

    #[test]
    fn the_identity_payload_carries_no_secret_or_command() {
        let payload = identity_payload(&RuntimeIdentity {
            effect_id: "effect".to_owned(),
            agent_run_id: "run".to_owned(),
            project_id: "project".to_owned(),
            issue_id: "issue".to_owned(),
            scope: "ticket".to_owned(),
            provider: "claude".to_owned(),
            target_kind: "work_item".to_owned(),
            target_id: "target".to_owned(),
            state: "prepared".to_owned(),
            attempt_count: 3,
        });
        let object = payload.as_object().expect("an object payload");
        assert_eq!(object.len(), 8);
        for forbidden in [
            "prompt",
            "command",
            "cwd",
            "env",
            "token",
            "api_key",
            "state",
            "attempt_count",
        ] {
            assert!(!object.contains_key(forbidden), "leaked {forbidden}");
        }
    }
}
